### PR TITLE
WIP: tiled renderer

### PR DESCRIPTION
**Pull request by MartinBuessemeyer**
**Description:**

This PR adds a TiledCameraGenerator-Class that takes a camera as an argument as well as a tile size and makes it possible to iterate over the provided camera in sub-tiles according to the provided tile size.  Currently a linear iterator is the only type of iteration supported for now.
The TiledCameraGenerator-Class has methods that return the tiles in the x and y dimension as well as the total number of tiles to support iterating over all tiles / sub-frustums via two different methods. The TiledCameraGenerator-Class supports three Iteration Algorithms to traverse the tiles:
Hilbert-Curve, Z-Curve and ScanLine.
In order to show that the tiled renderer is working, an example for the tiled renderer has been added to the webgl-operate website. It is a copy of the cube-example and renders the cube in original in the upper left quadrant. The other quadrants are rendered with the TileCameraGenerator-Class with the different Iteration Algorithms.

**Steps to test:**
Launch the webgl-operate server and view the tiled-renderer-example inside a Browser.

**Issues:**

    #186 - could be the definition for this task

### CODE REVIEW

**Comment by MartinBuessemeyer** [comment]
@cgcostume All bugs should be fixed:
The overdraw from the HilbertCurve is caused by the fact that the last Tile row is overlapping with the other quadrants. Even though these tiles need to be rendered. Thus I added gl.sissor like w.scheibel suggested.
The last row in the z-curve was not drawn since the number of tiles was no power of to. I changed the algorithm. Thus it should be able to handle non power of two numbers.
I reviewed the class comments and updated them.
I did not manage the write the tests but I will do it tomorrow.

tl:dr
I fixed the bugs and the tiled renderer should be merge-ready. I will add the tests tomorrow.

File: examples/tiled-rendering-example.ts [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+
+        this._intermediateFBO = new Framebuffer(this._context, 'IntermediateFBO');
+        this._blit = new BlitPass(this._context);
+        this._blit.initialize();
+        this._blit.readBuffer = gl2facade.COLOR_ATTACHMENT0;
+        this._blit.drawBuffer = gl.BACK;
+        this._blit.target = this._defaultFBO;
+        this._blit.framebuffer = this._intermediateFBO;
+
+        if (!this._intermediateFBO.initialized) {
```
Comment (scheibel) [inline examples/tiled-rendering-example.ts]: This code is within the `onInitialize` method, where the `_intermediateFBO` is set to a new instance without any condition.
I therefore suppose the FBO is always uninitialized.

File: examples/triangle-example.ts [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        this._texture.wrap(gl.REPEAT, gl.REPEAT);
+        this._texture.filter(gl.LINEAR, gl.LINEAR_MIPMAP_LINEAR);
+        this._texture.maxAnisotropy(Texture2D.MAX_ANISOTROPY);
+
+        this._texture.fetch('./data/triangle-texture.webp', false).then(() => {
+            const gl = context.gl;
+
+            this._program.bind();
+            gl.uniform1i(this._program.uniform('u_textured'), true);
+            console.log('enabled texture.');
```
Comment (scheibel) [inline examples/triangle-example.ts]: Please remove debug output or at least hide it behind certain verbosity levels (see https://github.com/cginternals/webgl-operate/blob/master/source/auxiliaries.ts#L89).

File: package.json [hunk 3]
```diff
@@ -33,7 +33,7 @@
         "@types/sinon": "~7.0.9",
         "chai": "~4.2.0",
         "chai-spies": "~1.0.0",
-        "coveralls": "~3.0.2",
+        "coveralls": "^3.0.6",
```
Comment (scheibel) [inline package.json]: I think I always mix the semantics of `~` and `^`, but is this change an intended fix or a deviation from our *explicit version* policy?

File: source/camera.ts [hunk 4]
```diff
@@ -365,4 +392,37 @@ export class Camera {
...
+     */
+    public copy(): Camera {
+        const copiedCamera = new Camera(this.eye, this.center, this.up);
+        copiedCamera.altered = this.altered;
+        copiedCamera.fovy = this.fovy;
+        copiedCamera.near = this.near;
+        copiedCamera.far = this.far;
+        copiedCamera.viewport = [this.viewport[0], this.viewport[1]];
+        copiedCamera.aspect = this.aspect;
+        copiedCamera.postViewProjection = this.postViewProjection;
```
Comment (scheibel) [inline source/camera.ts]: This matrix is passed by reference. As the camera is referred to as a *copied* camera, I assume that the internal projection matrix has to be copied, too.

Comment (scheibel) [reply source/camera.ts]: While reviewing the next method (which actually copies the projection matrix), I imagined the following implementation for `copy`:
```typescript
const copiedCamera = new Camera(/* ... */);
this.copyAllValues(copiedCamera);
return copiedCamera;
```

File: source/tilecameragenerator.ts [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+    protected _valid: boolean;
+    protected _currentOffset: [number, number] = [0, 0];
+    protected _iterationAlgorithmIndices: [number, number][] = [];
+
+    /**
+     * Converts the tile index from the selected Algorithm to table indices.
+     * @returns - The converted tile index.
+     */
+    protected tileIndexToTableIndices(): [number, number] {
+        if (this.algorithm === TileCameraGenerator.IterationAlgorithm.ScanLine) {
```
Comment (scheibel) [inline source/tilecameragenerator.ts]: Depending on the iteration algorithm, you chose different approaches:
1. In-line computation (scan line)
2. Precomputation and lookup (hilbert curve, z curve)
I suggest to convert all variations to precomputed `_iterationAlgorithmIndices` as it decouples generation logic and access logic. This results in significantly reduced code for access and a clear structure for generation.
One down side: the memory footprint of scan line algorithm slightly increases.

What seems to generally hold is that the contents on `_iterationAlgorithmIndices` solely rely on the algorithm and the width and height of the raster. As they usually don't match with the number of fragments of the canvas, we may even pre-compute the raster for several sizes.

File: source/tilecameragenerator.ts [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+            return [0, 0];
+        }
+    }
+
+    /**
+     * Fills the iterationAlgorithmIndices with the table indices
+     * from the ZCurve-Iteration-Algorithm.
+     */
+    protected fillIterationIndicesWithZCurve(): void {
+        this._iterationAlgorithmIndices = [];
```
Comment (scheibel) [inline source/tilecameragenerator.ts]: I suggest to don't use the `push` interface of an array but initialize the length up-front and assign the values instead of pushing them at the end of the array.
If the final number of entries is not known up-front, compute the upper bound and resize the array down afterwards.

File: source/tilecameragenerator.ts [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+        ++this.tile;
+        this.update();
+        return true;
+    }
+
+    /**
+     * Returns if tiles still need to be rendered.
+     */
+    public hasNextTile(): boolean {
+        if (this.tile < this.numberOfTiles() - 1 && this.tile >= 0) {
```
Comment (scheibel) [inline source/tilecameragenerator.ts]: Please use the more concise form `return this.tile < this.numberOfTiles() - 1 && this.tile >= 0;`

File: source/tilecameragenerator.ts [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+        this._currentOffset[0] = 0;
+        this._currentOffset[1] = 0;
+    }
+
+    /**
+     * Reassigns all values from the source camera to the tile camera.
+     * Should be called when the source camera is altered.
+     */
+    public updateCameraProperties(): void {
+        this.sourceCamera.copyAllValues(this.camera);
```
Comment (scheibel) [inline source/tilecameragenerator.ts]: This indicated that `sourceCamera` and `camera` should always hold similiar or even same values. Why do we still use a copy?

File: source/tilecameragenerator.ts [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+            this._camera = camera.copy();
+            this._valid = false;
+        }
+    }
+
+    /**
+     * Returns the current tile index.
+     * @returns - The current tile index.
+     */
+    get tile(): number {
```
Comment (scheibel) [inline source/tilecameragenerator.ts]: The name of the getter suggests that the current tile (whatever the semantics may be) is returned. However, the comment hints that the *current tile index* is returned instead. Why not name the getter `currentTileIndex`?

File: source/tilecameragenerator.ts [hunk 10]
```diff
@@ -0,0 +1,491 @@
...
+        return this._tile;
+    }
+
+    /**
+     * Sets the current tile index.
+     * @param index - The new index.
+     */
+    set tile(index: number) {
```
Comment (scheibel) [inline source/tilecameragenerator.ts]: Same here.

File: source/tilecameragenerator.ts [hunk 11]
```diff
@@ -0,0 +1,491 @@
...
+        }
+    }
+
+    /**
+     * Returns the size of the original viewport
+     * which should be divided in tiles based on the tile size.
+     * If the viewport has not been set, it returns [-1, -1].
+     * @returns - Size of the Viewport.
+     */
+    get sourceViewPort(): [number, number] {
```
Comment (scheibel) [inline source/tilecameragenerator.ts]: We use *viewport* as a single word. The correct camel-case name would be `sourceViewport`.

File: source/tilecameragenerator.ts [hunk 12]
```diff
@@ -0,0 +1,491 @@
...
+    }
+}
+
+/**
+ * The enum that is used to select one of the different the IterationAlgorithms.
+ */
+export namespace TileCameraGenerator {
+
+    export enum IterationAlgorithm {
+        ScanLine = 'scanline',
```
Comment (scheibel) [inline source/tilecameragenerator.ts]: Comments are generally placed above their target.

File: examples/tiled-rendering-example.ts (hunk 1 above)
Comment (MartinBuessemeyer) [reply examples/tiled-rendering-example.ts]: I guess you want to tell me that the if-statement is not needed. At least that would make sense to me.

File: package.json (hunk 3 above)
Comment (MartinBuessemeyer) [reply package.json]: I did not change the file. I guess it was vs-code. Shall I revert the changes?

File: source/camera.ts (hunk 4 above)
Comment (MartinBuessemeyer) [reply source/camera.ts]: Thanks there was definitely some code duplication.

File: source/tilecameragenerator.ts (hunk 8 above)
Comment (MartinBuessemeyer) [reply source/tilecameragenerator.ts]: Since we do not want to manipulate the camera provided by the user (```sourceCamera```). We change the ```postViewProjection``` on our camera (```camera```) to render the current tile, which would alter the users camera.
And yes it is true that the the only difference between these camera is the ```postViewProjection```.

File: source/tilecameragenerator.ts (hunk 9 above)
Comment (MartinBuessemeyer) [reply source/tilecameragenerator.ts]: In my opinion the name ```currentTileIndex``` is a better name. But cgcostume wanted it be called ```tile```.

File: source/tilecameragenerator.ts (hunk 10 above)
Comment (MartinBuessemeyer) [reply source/tilecameragenerator.ts]: Please see above.

**Comment by MartinBuessemeyer** [comment]
@scheibel Thanks for the comments. I implemented all of your requested changes except the `tile`-renaming.
Additionally I finished writing some basic tests for the iteration algorithms. Shall I write more tests?

File: examples/tiled-rendering-example.ts (hunk 1 above)
Comment (scheibel) [reply examples/tiled-rendering-example.ts]: Basically, yes. The other option is that I was missing something, thus, my explicit reasoning.

File: package.json (hunk 3 above)
Comment (scheibel) [reply package.json]: @cgcostume What is the desired approach?
